### PR TITLE
[stable/redis-ha] use master group name not used in haproxy tcp-check…

### LINKED ISSUE #17868: [stable/redis-ha] master group name not used in haproxy tcp-check
**Describe the bug**

When using another value than mymaster for masterGroupName, haproxy doesnt find any master.

**Version of Helm and Kubernetes**:

**Which chart**:
stable/redis-ha

**What happened**:

we have a lot of redis-ha chart installed in our cluster.
under certains circumstances one sentinel from a redis-ha release will join another sentinel from another release.
we dont found the issue yet but an option to us is to use separate mymaster groups, even if this situation happens, this will not break the sentinel quorum

**What you expected to happen**:

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by cschockaert**
… (#17868)

Signed-off-by: Clément Schockaert <clement.schockaert@outlook.com>

#### Which issue this PR fixes
  - fixes #17868

#### Checklist
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

File: stable/redis-ha/templates/_configs.tpl
Comment (zanhsieh) [inline stable/redis-ha/templates/_configs.tpl]: ==> Linting stable/redis-ha
[ERROR] templates/: render error in "redis-ha/templates/redis-haproxy-deployment.yaml": template: redis-ha/templates/redis-haproxy-deployment.yaml:27:35: executing "redis-ha/templates/redis-haproxy-deployment.yaml" at <include "config-hapr...>: error calling include: template: redis-ha/templates/_configs.tpl:166:66: executing "config-haproxy.cfg" at <.Values.redis.master...>: can't evaluate field Values in type int

Could you fix this?

Comment (cschockaert) [reply stable/redis-ha/templates/_configs.tpl]: Got it fixed, thanks

**Comment by cschockaert** [comment]
@zaneclaes seems good to me

**Comment by cschockaert** [comment]
We are running into another issue, since we are using this chart in multiples release and we want to be more flexible, we need to be able to tpl against the masterGroupName or at least use something like :

```
{{- define "redis-ha.masterGroupName" -}}
{{- if .Values.redis.masterGroupName -}}
{{ .Values.redis.masterGroupName }} # or even use tpl here ?
{{- else -}}
{{ .Release.Name }}-mymaster
{{- end -}}
{{- end -}}
```

is it ok to you to push a change like that?

**Comment by ssalaues** [comment]
> We are running into another issue, since we are using this chart in multiples release and we want to be more flexible, we need to be able to tpl against the masterGroupName or at least use something like :
> 
> ```
> {{- define "redis-ha.masterGroupName" -}}
> {{- if .Values.redis.masterGroupName -}}
> {{ .Values.redis.masterGroupName }} # or even use tpl here ?
> {{- else -}}
> {{ .Release.Name }}-mymaster
> {{- end -}}
> {{- end -}}
> ```
> 
> is it ok to you to push a change like that?

@cschockaert  Your current changes in the PR look good and very necessary. However, just so I'm sure I understand what you're asking above, you want to add a tpl function in the case the masterGroupName is not defined? If so, I think this is okay. @DandyDeveloper any thoughts on this?

**Comment by DandyDeveloper** [comment]
@cschockaert @ssalaues This looks like a good change, but I want to play devils advocate to understand the justification of the group name change. 

Would the more appropriate thing for multiple releases, be to use `--set` in the helm install command? This is what I do. 

If Salim is right and you want to empty the master and use a tpl for generating it, I'm not sure this is a good change, as it isn't very clear to people using the chart and can probably be a pain in the ass to debug for people using this.

**Comment by cschockaert** [comment]
@DandyDeveloper you are right, after all i think that emptying master name should just be prohibed.
Well all we need here is the capability of templating the master group name, i would suggest a change like that every where we need master group name:

{{ tpl .Values.redis.masterGroupName $ }}

or somehting in helpers like that:
```
{{- define "redis-ha.masterGroupName" -}}
{{- $masterGroupName := tpl ( .Values.redis.masterGroupName | default "mymaster") $ -}}
{{- $validMasterGroupName := regexMatch "^[A-Za-z0-9_-]+$" $masterGroupName -}}
{{- if $validMasterGroupName -}}
{{ $masterGroupName | lower }}
{{- else -}}
{{ required "A valid .Values.redis.masterGroupName entry is required (matching ^[A-Za-z0-9_-]+$)" ""}}
{{- end -}}
{{- end -}}
```

masterGroupName will be check against a regexp
(Note: master name should not include special characters or spaces.
# The valid charset is A-z 0-9 and the three characters ".-_".)

empty masterGroupName will be mymaster
masterGroupName can be templated inside values.yaml

**Comment by cschockaert** [comment]
To be clear why we need that tpl feature, its because we got a lot of redis-ha chart installed in our preemptible instance, and under certains circumstances sentinel of another redis-ha release joined together and leading to uncomprehensible situation where we got many redis at prior unknown of each other (differents redis-ha charts) working (so not well working) together

**Comment by DandyDeveloper** [comment]
@cschockaert I actually agree that a validation template function would be ideal. That's a great suggestion to prevent people scratching their heads why their instances are not running. 

> under certains circumstances sentinel of another redis-ha release joined together 

This is strange behaviour... Are you using namespaces for the different releases? I am not sure how this can happen when the announce addresses are explicit to the release name. Considering that we explicitly use the IP for those announce addresses as well, I am not sure how this kind of behavior happens.

**Comment by cschockaert** [comment]
We are using differents namespace for redis-ha and fullnameOverride.

for example with fullname override = product-api-redis-ha

so we can have theses svc in namespace one:

```
product-api-redis-ha
product-api-redis-ha-announce-0
product-api-redis-ha-announce-1
product-api-redis-ha-announce-2
product-api-redis-ha-haproxy
```

and the same one in another namespace.

I saw in config that we resolve ip against fullname but we dont prefix any namespace in dns
ex : 
{{ template "redis-ha.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local
product-api-redis-ha.namespacea.svc.cluster.local

but without adding {{ .Release.Namespace }}.svc.cluster.local, kubedns will normally resolve in current namespace.

Another thing i saw is that our mixed releases are not mixed with the same fullnameOverride. it's totally random..

**Review by zanhsieh — APPROVED** [review]

File: stable/redis-ha/templates/_helpers.tpl
Comment (ssalaues) [inline stable/redis-ha/templates/_helpers.tpl:77]: According to the Redis Sentinel:
```
The valid charset is A-z 0-9 and the three characters ".-_".
```
I am not sure I understand what range you are trying to match here.

File: stable/redis-ha/templates/_helpers.tpl
Comment (ssalaues) [inline stable/redis-ha/templates/_helpers.tpl]: This shouldn't be lowered.

**Review by ssalaues — CHANGES_REQUESTED** [review]

**Comment by ssalaues** [comment]
> We are using differents namespace for redis-ha and fullnameOverride.
> 
> for example with fullname override = product-api-redis-ha
> ...
> I saw in config that we resolve ip against fullname but we dont prefix any namespace in dns
> ex :
> {{ template "redis-ha.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local
> product-api-redis-ha.namespacea.svc.cluster.local
> 
> but without adding {{ .Release.Namespace }}.svc.cluster.local, kubedns will normally resolve in current namespace.
> 
> Another thing i saw is that our mixed releases are not mixed with the same fullnameOverride. it's totally random..

So I understand that you are using essentially the same service names in your various namespaces, but the point of the namespace is exactly for this reason. There should be no reason that I can see why your redis quorum in `namespace-a` can attach to another redis in `namespace-b` since (exactly as you said) without adding the ` {{ .Release.Namespace }}.svc.cluster.local` it will always resolve within the same namespace therefore preventing the redis/sentinels joining across namespaces. So the question is, why are your redis instances attaching across namespaces?

**Comment by cschockaert** [comment]
Well, i don't know how it happened, since i use isolation with differents group name it did not happen again.

But we are using istio and preemtible instance.
There is a lot of crashloopbackoff coz of istio slowing down startup of pods.
Perhaps one pod got rescheduled with another IP and the old IP was used by another redis/sentinel instance, leading to other sentinel monitoring this pod by ip allow him to join the redis cluster.
I mean, since redis.conf and sentinel.conf is populated by IP and not dns, this can happen

**Comment by cschockaert** [comment]
We have a situation too (caused by slow startup and K8S killing pod) where for example a cluster of 3 instance, we got 

redis-0 slave and replica of redis-2
redis-1 slave and replica of redis-0
redis-2 master 

this happened after a rollingupdate restarting redis-2 then 1 then 0

perhaps due to a race condition in config-init choosing index 0 as master, and a sentinel quorum relecting a new master.

i have to investiguate about that too...

File: stable/redis-ha/templates/_helpers.tpl
Comment (cschockaert) [reply stable/redis-ha/templates/_helpers.tpl:77]: `\w Matches any letter, digit or underscore. Equivalent to [a-zA-Z0-9_].`

i added - and . to the allowed chars, seems fine to me but i can change to regexp to not use \w if you want,  which can be confusing i know

File: stable/redis-ha/templates/_helpers.tpl
Comment (cschockaert) [reply stable/redis-ha/templates/_helpers.tpl]: seems right, i misunderstood the documentation (uppercase is allowed) fixed

**Comment by cschockaert** [comment]
@ssalaues @DandyDeveloper seems to me that all is resolved around master name customization, can we merge it?
thanks :)

**Comment by cschockaert** [comment]
Hello, just updated this MR, @ssalaues @DandyDeveloper

**Comment by cschockaert** [comment]
@ssalaues @DandyDeveloper someone here?

**Comment by DandyDeveloper** [comment]
@ssalaues Simply because you've been so heavily involved in the discussion. Can you approve this if it looks good to you? 

If you want me to get involved, let me know and I'll weigh in on the discussion. I don't want to LGTM @cschockaert just incase Salim has some thoughts.

**Comment by cschockaert** [comment]
Nice, just updated this MR to get rid of conflict (chart version)

**Pull request merged**
